cmd/jaeger/internal/integration/trace_writer.go

```diff
@@ -0,0 +1,134 @@
+
+package integration
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"time"
+
+	"go.opentelemetry.io/collector/component/componenttest"
+	"go.opentelemetry.io/collector/config/configtls"
+	"go.opentelemetry.io/collector/exporter"
+	"go.opentelemetry.io/collector/exporter/exportertest"
+	"go.opentelemetry.io/collector/exporter/otlpexporter"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"go.uber.org/zap"
+
+	"github.com/jaegertracing/jaeger/internal/jptrace"
+	"github.com/jaegertracing/jaeger/storage_v2/tracestore"
+)
+
+var (
+	_ tracestore.Writer = (*traceWriter)(nil)
+	_ io.Closer         = (*traceWriter)(nil)
+)
+
+// traceWriter utilizes the OTLP exporter to send span data to the Jaeger-v2 receiver
+type traceWriter struct {
+	logger   *zap.Logger
+	exporter exporter.Traces
+}
+
+func createTraceWriter(logger *zap.Logger, port int) (*traceWriter, error) {
+	logger.Info("Creating the trace writer", zap.Int("port", port))
+
+	factory := otlpexporter.NewFactory()
+	cfg := factory.CreateDefaultConfig().(*otlpexporter.Config)
+	cfg.Endpoint = fmt.Sprintf("localhost:%d", port)
+	cfg.Timeout = 30 * time.Second
+	cfg.RetryConfig.Enabled = false
+	cfg.QueueConfig.Enabled = false
+	cfg.TLSSetting = configtls.ClientConfig{
+		Insecure: true,
+	}
+
+	set := exportertest.NewNopSettings()
+	set.Logger = logger
+
+	exp, err := factory.CreateTraces(context.Background(), set, cfg)
+	if err != nil {
+		return nil, err
+	}
+	if err := exp.Start(context.Background(), componenttest.NewNopHost()); err != nil {
+		return nil, err
+	}
+
+	return &traceWriter{
+		logger:   logger,
+		exporter: exp,
+	}, nil
+}
+
+func (w *traceWriter) Close() error {
+	w.logger.Info("Closing the trace writer")
+	return w.exporter.Shutdown(context.Background())
+}
+
+func (w *traceWriter) WriteTraces(ctx context.Context, td ptrace.Traces) error {
+	var err error
+	maxChunkSize := 35 // max chunk size otel kafka export can handle safely.
+	if td.SpanCount() > maxChunkSize {
+		w.logger.Info("Chunking traces")
+		err = w.writeTraceInChunks(ctx, td, maxChunkSize)
+		w.logger.Info("Finished writing chunks")
+	} else {
+		err = w.exporter.ConsumeTraces(ctx, td)
+	}
+	return err
+}
+
+// writeTraceChunks splits td into before writing
+
+// td span count should be greater than maxChunkSize
+func (w *traceWriter) writeTraceInChunks(ctx context.Context, td ptrace.Traces, maxChunkSize int) error {
+	var err error
+
+	currentChunk := ptrace.NewTraces()
+	currentResourceIndex := -1
+	currentScopeIndex := -1
+	spanCount := 0
+
+	jptrace.SpanIter(td)(func(pos jptrace.SpanIterPos, span ptrace.Span) bool {
+		var (
+			scope    ptrace.ScopeSpans
+			resource ptrace.ResourceSpans
+		)
+		// Create a new chunk if the current span count reaches MaxChunkSize
+		if spanCount == maxChunkSize {
+			err = w.exporter.ConsumeTraces(ctx, currentChunk)
+			currentChunk = ptrace.NewTraces()
+			spanCount = 0
+		}
+
+		if currentChunk.ResourceSpans().Len() == 0 || currentResourceIndex != pos.ResourceIndex {
```

Comment:
is the 2nd clause not sufficient?
```suggestion
		if currentResourceIndex != pos.ResourceIndex {
```

Reply:
by resetting the indices, it will be sufficient😱🤗✔✔